File: Programação orientada a objetos/POO-241 - Trab 2 - Pedro_Quedevez/Recredenciamento.cpp
#include <unordered_map>
#include <iomanip>
#include "StringUtils.hpp"
#include "Regras.hpp"
#include "Recredenciamento.hpp"
#include "ItemListaPublicacoes.hpp"

using namespace cpp_util;

Recredenciamento::Recredenciamento(string nomeDocente, double pontuacao, string recrendenciado)
{
    this->nomeDocente = nomeDocente;
    this->pontuacao = pontuacao;
    this->recrendenciado = recrendenciado;
}

vector<Recredenciamento> Recredenciamento::constroiListaRecredenciamento(vector<Docente> docentes, vector<Regras> regras, int anoAtual)
{
    vector<Recredenciamento> recredenciamentos;
    vector<float> vetorPontos = constroiVetorPontos(regras, anoAtual);

    Regras regraAtual = Regras::getRegraAtual(regras, anoAtual);
    int pontuacaoMinima = regraAtual.getPontuacaoMin();

    for (Docente docente : docentes)
    {
        float pontuacao = calculaPontuacao(vetorPontos, docente, regraAtual.getQtdeAnosPontos(), anoAtual);
        string evento = verificaEventoRecredenciamento(docente, anoAtual);
        bool publicacaoMinima = verificaPubMin(regras, docente, anoAtual);

        if (evento != "")
        {
            Recredenciamento r = Recredenciamento(docente.getNome(), pontuacao, evento);
            recredenciamentos.push_back(r);
        }
        else if (pontuacao >= pontuacaoMinima && publicacaoMinima && evento == "")
        {
            Recredenciamento r = Recredenciamento(docente.getNome(), pontuacao, "Sim");
            recredenciamentos.push_back(r);
        }
        else
        {
            Recredenciamento r = Recredenciamento(docente.getNome(), pontuacao, "Não");
            recredenciamentos.push_back(r);
        }
    }

    return recredenciamentos;
}

bool Recredenciamento::verificaPubMin(vector<Regras> regras, Docente docente, int anoAtual)
{
    Regras regraAtual = Regras::getRegraAtual(regras, anoAtual);
    vector<string> qualis2 = regraAtual.getQualis2();
    vector<int> quantidadesMinimas = regraAtual.getQuantidadesMinimas();

    int quantMin = quantidadesMinimas[0];
    string qualisMin = qualis2[qualis2.size() - 1];

    vector<Publicacao> publicacoes = docente.getPublicacoes();

    for (Publicacao p : publicacoes)
    {
        Veiculo v = p.getVeiculo();
        v.getTipo();
        if ((ordemQualisMap[p.getQuali()] <= ordemQualisMap[qualisMin]) && (v.getTipo() == "P" || v.getTipo() == "p"))
        {
            quantMin--;
        }
    }

    if (quantMin > 0)
    {
        return false;
    }
    else
    {
        return true;
    }
}

vector<float> Recredenciamento::constroiVetorPontos(vector<Regras> regras, int anoAtual)
{
    vector<float> vetorPontos(10, 0);

    Regras regraAtual = Regras::getRegraAtual(regras, anoAtual);
    vector<string> qualis1 = regraAtual.getQualis1();
    vector<string> pontos = regraAtual.getPontos();
    for (int i = 0; i < pontos.size(); i++)
    {
        pontos[i] = trocarVirgulaPorPonto(pontos[i]);
    }
    int qtdeAnosPontos = regraAtual.getQtdeAnosPontos();

    for (int i = 0; i < qualis1.size(); i++)
    {
        vetorPontos[ordemQualisMap[qualis1[i]]] = stof(pontos[i]); // pontos e qualis1 tem o mesmo tamanho sempre;

        for (int j = ordemQualisMap[qualis1[i]]; j < 10; j++) // preenche o vetor do ponto atual até o final com o mesmo valor;
        {
            vetorPontos[j] = stof(pontos[i]);
        }
    }

    return vetorPontos;
}

float Recredenciamento::calculaPontuacao(vector<float> &vetorPontos, Docente docente, int quantAnos, int anoAtual)
{
    float pontuacao = 0.0f;
    vector<Publicacao> publicacoes = docente.getPublicacoes();
    int limiteAno = anoAtual - quantAnos;

    for (Publicacao p : publicacoes)
    {
        if (p.getAno() >= limiteAno && p.getAno() < anoAtual)
        {
            pontuacao += vetorPontos[ordemQualisMap[p.getQuali()]];
        }
    }
    return pontuacao;
}

string Recredenciamento::verificaEventoRecredenciamento(Docente docente, int anoAtual)
{
    vector<Ocorrencia> ocorrenciasDocente = docente.getOcorrencias();
    for (Ocorrencia o : ocorrenciasDocente)
    {
        string evento = o.getEvento();
        if (trim(evento) == "Coordenador" || trim(evento) == "Licença Maternidade" || trim(evento) == "Bolsista CNPq")
        {
            return trim(evento);
        }
    }

    if (docente.idadeDocente(anoAtual) >= 60)
    {
        return "PPS";
    }

    if (docente.tempoIngresso(anoAtual) <= 2)
    {
        return "PPJ";
    }

    return "";
}

vector<Recredenciamento> Recredenciamento::ordenaListaRecredenciamento(vector<Recredenciamento> listaRecredenciamento)
{
    for (int i = 0; i < listaRecredenciamento.size(); i++)
    {
        for (int j = i + 1; j < listaRecredenciamento.size(); j++)
        {
            if (listaRecredenciamento[i].getNomeDocente() > listaRecredenciamento[j].getNomeDocente())
            {
                Recredenciamento aux = listaRecredenciamento[i];
                listaRecredenciamento[i] = listaRecredenciamento[j];
                listaRecredenciamento[j] = aux;
            }
        }
    }
    return listaRecredenciamento;
}

string Recredenciamento::getNomeDocente()
{
    return this->nomeDocente;
}

double Recredenciamento::getPontuacao()
{
    return this->pontuacao;
}

string Recredenciamento::getRecrendenciado()
{
    return this->recrendenciado;
}

void Recredenciamento::setNomeDocente(string nomeDocente)
{
    this->nomeDocente = nomeDocente;
}

void Recredenciamento::setPontuacao(double pontuacao)
{
    this->pontuacao = pontuacao;
}

void Recredenciamento::setRecrendenciado(string recrendenciado)
{
    this->recrendenciado = recrendenciado;
}

ostream &operator<<(ostream &o, Recredenciamento &r)
{
    stringstream pontuacaoAux;
    pontuacaoAux << fixed << setprecision(1) << r.getPontuacao();
    string pontuacao = trocarPontoPorVirgula(pontuacaoAux.str());

    o << r.getNomeDocente() << ";" << pontuacao << ";" << r.getRecrendenciado();
    return o;
}

File: Programação orientada a objetos/POO 241 - Trab 1 - Pedro Quedevez/Recredenciamento.java

import java.util.ArrayList;
import java.util.Comparator;

public class Recredenciamento {
    String nomeDocente;
    double pontuacao;
    String recredenciado;

    public enum ordemQualis {
        A1, A2, A3, A4, B1, B2, B3, B4, B5, C;
    }

    public static ArrayList<Recredenciamento> listaRecredenciamento(ArrayList<Docente> docentes,
            ArrayList<Regras> regras, ArrayList<Veiculo> veiculos, int anoAtual) {

        // obtem as informações da regra necessárias para o recredenciamento;
        Regras regraAtual = Regras.getRegraAtual(regras, anoAtual);
        int quantAnosPontos = regraAtual.getQtdeAnosPontos();
        Float vetorPontos[] = regraAtual.constroiVetorPontos();

        ArrayList<Recredenciamento> listaRecredenciamento = new ArrayList<>();

        for (Docente docente : docentes) {
            // obtem informações de recadastramento do docente para comparação;
            String recredenciado = Ocorrencia.EventoRecadastramento(docente.getOcorrencias(), anoAtual);
            Float pontuacao = Recredenciamento.calculaPontuacao(vetorPontos, docente, quantAnosPontos, anoAtual);

            int regraAtualQuantMin = regraAtual.getQuantidadesMinimas().get(0);
            String regraAtualQualisMin = regraAtual.getQualis1().get(regraAtual.getQualis1().size() - 1);

            // cria e adiciona o resultado para o docente;
            if (recredenciado.equals("Nao")) {
                if (docente.idadeDocente(anoAtual) >= 60) {
                    recredenciado = "PPS";
                } else if (docente.tempoIngresso(anoAtual) <= 2) {
                    recredenciado = "PPJ";
                } else if (pontuacao >= regraAtual.getPontuacaoMin()
                        && Recredenciamento.verificaPubMin(veiculos, docente,
                                anoAtual, regraAtual.getQtdeAnosArtigo(), regraAtualQuantMin, regraAtualQualisMin)) {
                    recredenciado = "Sim";
                } else {
                    recredenciado = "Não";
                }
            }

            listaRecredenciamento.add(new Recredenciamento(docente.getNome(), pontuacao,
                    recredenciado));

        }
        listaRecredenciamento.sort(new listaComparatorR());
        return listaRecredenciamento;
    }

    // auxilia na verificação de publicações mínimas;
    public static Boolean verificaPubMin(ArrayList<Veiculo> veiculos, Docente docente, int anoAtual,
            int quantAnosPontos, int quantPubMin,
            String QualiMin) {

        ArrayList<Publicacao> publicacoes = docente.getPublicacoes();
        int limiteAno = anoAtual - quantAnosPontos;
        int quantMinAux = quantPubMin;

        for (Publicacao publicacao : publicacoes) {
            Enum<ordemQualis> qualisEnumPub = ordemQualis.valueOf(publicacao.getQuali());
            Enum<ordemQualis> qualisEnumMin = ordemQualis.valueOf(QualiMin);
            Character tipoVeiculo = Veiculo.verificaTipoVeiculo(veiculos, publicacao.getSiglaVeiculo());

            if ((publicacao.getAno() >= limiteAno) && (qualisEnumPub.ordinal() <= qualisEnumMin.ordinal())
                    && (tipoVeiculo == 'P')) {
                quantMinAux--;
            }
        }

        if (quantMinAux <= 0) {
            return true;
        } else {
            return false;
        }

    }

    public static Float calculaPontuacao(Float vetorPontos[], Docente docente, int quantAnosPontos, int anoAtual) {
        Float pontuacao = 0.0f;
        ArrayList<Publicacao> publicacoes = docente.getPublicacoes();
        for (Publicacao publicacao : publicacoes) {
            String quali = publicacao.getQuali().trim();
            int anoPublicacao = publicacao.getAno();
            int limiteAno = anoAtual - quantAnosPontos;
            if (anoPublicacao >= limiteAno && anoPublicacao < anoAtual) {
                if (quali.equals("A1")) {
                    pontuacao += vetorPontos[0].floatValue();
                } else if (quali.equals("A2")) {
                    pontuacao += vetorPontos[1].floatValue();
                } else if (quali.equals("A3")) {
                    pontuacao += vetorPontos[2].floatValue();
                } else if (quali.equals("A4")) {
                    pontuacao += vetorPontos[3].floatValue();
                } else if (quali.equals("B1")) {
                    pontuacao += vetorPontos[4].floatValue();
                } else if (quali.equals("B2")) {
                    pontuacao += vetorPontos[5].floatValue();
                } else if (quali.equals("B3")) {
                    pontuacao += vetorPontos[6].floatValue();
                } else if (quali.equals("B4")) {
                    pontuacao += vetorPontos[7].floatValue();
                } else if (quali.equals("B5")) {
                    pontuacao += vetorPontos[8].floatValue();
                } else if (quali.equals("C")) {
                    pontuacao += vetorPontos[9].floatValue();
                }
            }
        }
        return pontuacao;
    }

    public Recredenciamento(String nomeDocente, double pontuacao, String recredenciado) {
        this.nomeDocente = nomeDocente;
        this.pontuacao = pontuacao;
        this.recredenciado = recredenciado;
    }

    @Override
    public String toString() {
        return nomeDocente + ";" + String.format("%.1f", pontuacao) + ";" + recredenciado;
    }

    public String getNomeDocente() {
        return nomeDocente;
    }

    public void setNomeDocente(String nomeDocente) {
        this.nomeDocente = nomeDocente;
    }

    public double getPontuacao() {
        return pontuacao;
    }

    public void setPontuacao(double pontuacao) {
        this.pontuacao = pontuacao;
    }

    public String getRecredenciado() {
        return recredenciado;
    }

    public void setRecredenciado(String recredenciado) {
        this.recredenciado = recredenciado;
    }

    public static class listaComparatorR implements Comparator<Recredenciamento> {
        @Override
        public int compare(Recredenciamento r1, Recredenciamento r2) {
            return r1.getNomeDocente().compareTo(r2.getNomeDocente());
        }
    }

}
